File: python/tests/wire/vbp/test_handlers.py
"""Tests for the vbp.wire.handlers module — all 23 mandatory opcodes have stubs."""
from __future__ import annotations

import os
import sys
import unittest

# Make sure we can import the vedadb package from the repo root.
HERE = os.path.dirname(os.path.abspath(__file__))
REPO_ROOT = os.path.abspath(os.path.join(HERE, "..", "..", ".."))
if REPO_ROOT not in sys.path:
    sys.path.insert(0, REPO_ROOT)

from vedadb.wire.vbp.handlers import (  # noqa: E402
    HANDLERS,
    assert_all_mandatory_handlers_registered,
)
from vedadb.wire.vbp.opcodes import (  # noqa: E402
    MANDATORY_OPCODES,
    OP_AUTH_CHALLENGE,
    OP_AUTH_OK,
    OP_AUTH_RESPONSE,
    OP_BEGIN,
    OP_BIND,
    OP_CANCEL_QUERY,
    OP_CLIENT_HELLO,
    OP_CLOSE,
    OP_COMMAND_COMPLETE,
    OP_COMMIT,
    OP_COPY_DONE,
    OP_COPY_FAIL,
    OP_COPY_IN,
    OP_DATA_CHUNK,
    OP_ERROR,
    OP_EXT_QUERY,
    OP_PARSE,
    OP_PING,
    OP_PONG,
    OP_QUERY,
    OP_ROLLBACK,
    OP_ROWS_FINISHED,
    OP_SERVER_READY,
)


class TestAllMandatoryOpcodes(unittest.TestCase):
    def test_23_mandatory_opcodes(self):
        self.assertEqual(len(MANDATORY_OPCODES), 23)

    def test_assert_raises_nothing_when_complete(self):
        # Should not raise.
        assert_all_mandatory_handlers_registered()

    def test_handler_for_every_mandatory_opcode(self):
        for op in MANDATORY_OPCODES:
            self.assertIn(op, HANDLERS, f"opcode 0x{op:02x} has no handler")

    def test_handlers_are_callable(self):
        for op, fn in HANDLERS.items():
            self.assertTrue(callable(fn), f"opcode 0x{op:02x} handler is not callable")


class TestRealHandlers(unittest.TestCase):
    def test_ping_handler_echoes_nonce(self):
        from vedadb.wire.vbp.handlers import handle_ping
        nonce = b"\x01\x02\x03\x04\x05\x06\x07\x08"
        replies = handle_ping(None, nonce)
        self.assertEqual(len(replies), 1)
        self.assertEqual(replies[0].op, OP_PONG)
        self.assertEqual(replies[0].body, nonce)

    def test_query_handler_select_1(self):
        from vedadb.wire.vbp.handlers import handle_query
        sql = b"SELECT 1"
        body = (
            # u32 query_id
            b"\x00\x00\x00\x00"
            # u32 text_len + text
            + (len(sql)).to_bytes(4, "little") + sql
            # u16 param_count
            + b"\x00\x00"
        )
        replies = handle_query(None, body)
        ops = [r.op for r in replies]
        self.assertIn(OP_DATA_CHUNK, ops)
        self.assertIn(OP_ROWS_FINISHED, ops)
        self.assertIn(OP_COMMAND_COMPLETE, ops)

    def test_client_hello_handler(self):
        from vedadb.wire.vbp.handlers import handle_client_hello
        import struct
        # u16 proto=1, u16 flags=0, u32 un_len=4, "user", u32 db_len=0, u8 ak=0, u32 aid_len=0
        body = struct.pack("<HHI4sI", 1, 0, 4, b"user", 0) + b"\x00" + struct.pack("<I", 0)
        replies = handle_client_hello(None, body)
        ops = [r.op for r in replies]
        self.assertIn(OP_SERVER_READY, ops)
        self.assertIn(OP_AUTH_OK, ops)

    def test_client_hello_wrong_proto_returns_error(self):
        from vedadb.wire.vbp.handlers import handle_client_hello
        import struct
        body = struct.pack("<HHI4sI", 99, 0, 4, b"user", 0) + b"\x00" + struct.pack("<I", 0)
        replies = handle_client_hello(None, body)
        self.assertEqual(replies[0].op, OP_ERROR)


if __name__ == "__main__":
    unittest.main()

File: node/tests/wire/vbp/test_handlers.js
'use strict';

const { test } = require('node:test');
const assert = require('node:assert');
const {
  HANDLERS, assertAllMandatoryHandlersRegistered,
  errFrame, dataChunkInt, rowsFinished, commandComplete,
} = require('../../../src/wire/vbp/handlers');
const {
  MANDATORY_OPCODES,
  OP_CLIENT_HELLO, OP_QUERY, OP_PING, OP_PONG, OP_AUTH_OK, OP_AUTH_RESPONSE,
  OP_BEGIN, OP_COMMIT, OP_ROLLBACK, OP_CLOSE, OP_ERROR, OP_DATA_CHUNK,
  OP_ROWS_FINISHED, OP_COMMAND_COMPLETE, OP_EXT_QUERY, OP_COPY_IN,
} = require('../../../src/wire/vbp/opcodes');

test('handlers: all 23 mandatory opcodes have a handler', () => {
  assertAllMandatoryHandlersRegistered();
  for (const op of MANDATORY_OPCODES) {
    assert.ok(typeof HANDLERS[op] === 'function', `missing handler for 0x${op.toString(16)}`);
  }
});

test('handlers: client_hello returns SERVER_READY + AUTH_OK', () => {
  const replies = HANDLERS[OP_CLIENT_HELLO](null, Buffer.alloc(0));
  assert.strictEqual(replies.length, 2);
  assert.strictEqual(replies[0].op, 0x02);
  assert.strictEqual(replies[1].op, 0x05);
});

test('handlers: query stub returns 1-row DATA_CHUNK + ROWS_FINISHED + COMMAND_COMPLETE', () => {
  const replies = HANDLERS[OP_QUERY](null, Buffer.alloc(0));
  assert.strictEqual(replies.length, 3);
  assert.strictEqual(replies[0].op, 0x0A);
  assert.strictEqual(replies[1].op, 0x0B);
  assert.strictEqual(replies[2].op, 0x0C);
});

test('handlers: ping echoes nonce in PONG', () => {
  const nonce = Buffer.from('12345678', 'hex');
  const replies = HANDLERS[OP_PING](null, nonce);
  assert.strictEqual(replies.length, 1);
  assert.strictEqual(replies[0].op, 0x17);
  assert.deepStrictEqual(replies[0].body, nonce);
});

test('handlers: close returns empty list', () => {
  const replies = HANDLERS[OP_CLOSE](null, Buffer.alloc(0));
  assert.strictEqual(replies.length, 0);
});

test('handlers: begin returns COMMAND_COMPLETE with in_txn=1', () => {
  const replies = HANDLERS[OP_BEGIN](null, Buffer.alloc(0));
  assert.strictEqual(replies.length, 1);
  assert.strictEqual(replies[0].op, 0x0C);
  assert.strictEqual(replies[0].body[0], 1);
});

test('handlers: commit/rollback return COMMAND_COMPLETE with in_txn=0', () => {
  for (const op of [OP_COMMIT, OP_ROLLBACK]) {
    const replies = HANDLERS[op](null, Buffer.alloc(0));
    assert.strictEqual(replies[0].op, 0x0C);
    assert.strictEqual(replies[0].body[0], 0);
  }
});

test('handlers: ext_query is stubbed (0A000)', () => {
  const replies = HANDLERS[OP_EXT_QUERY](null, Buffer.alloc(0));
  assert.strictEqual(replies.length, 1);
  assert.strictEqual(replies[0].op, 0x0D);
  // sqlstate = '0A000'
  assert.strictEqual(replies[0].body.subarray(0, 5).toString('ascii'), '0A000');
});

test('handlers: copy_in is stubbed (0A000)', () => {
  const replies = HANDLERS[OP_COPY_IN](null, Buffer.alloc(0));
  assert.strictEqual(replies.length, 1);
  assert.strictEqual(replies[0].op, 0x0D);
});

test('handlers: server-to-client opcodes reject', () => {
  for (const op of [OP_DATA_CHUNK, OP_ROWS_FINISHED, OP_COMMAND_COMPLETE, OP_ERROR, OP_PONG]) {
    const replies = HANDLERS[op](null, Buffer.alloc(0));
    assert.strictEqual(replies.length, 1);
    assert.strictEqual(replies[0].op, 0x0D);
  }
});

test('handlers: errFrame builder', () => {
  const f = errFrame(0x01, '08P01', 'oops');
  assert.strictEqual(f.op, 0x0D);
  assert.strictEqual(f.seq, 0x01);
  assert.strictEqual(f.body.subarray(0, 5).toString('ascii'), '08P01');
});

test('handlers: dataChunkInt builder shape', () => {
  const f = dataChunkInt(0x01, 42);
  assert.strictEqual(f.op, 0x0A);
  // chunk_id(4) + row_count(4) + col_count(2) + type_id(2) + bitmap(1) + value(4) = 17
  assert.strictEqual(f.body.length, 17);
  assert.strictEqual(f.body.readInt32LE(13), 42);
});

test('handlers: rowsFinished builder', () => {
  const f = rowsFinished(0x01, 7, 'SELECT 1');
  assert.strictEqual(f.op, 0x0B);
  assert.strictEqual(Number(f.body.readBigUInt64LE(0)), 7);
});

test('handlers: commandComplete builder', () => {
  const f = commandComplete(0x01, 0);
  assert.strictEqual(f.op, 0x0C);
  assert.strictEqual(f.body[0], 0);
});
